fix(CA2023: Adds validation against invalid braces in logger message templates)

Kritner opened the pull request:
* Fixes #7285

<!--

Make sure you have read the contribution guidelines: 
- https://learn.microsoft.com/contribute/dotnet/dotnet-contribute-code-analysis#contribute-docs-for-caxxxx-rules
- https://github.com/dotnet/roslyn-analyzers/blob/main/GuidelinesForNewRules.md

If your Pull Request is doing one of the following:

- Adding a new diagnostic analyzer or a code fix
- Adding or updating resource strings used by analyzers and code fixes
- Updating analyzer package versions in [Versions.props](../eng/Versions.props)

Then, make sure to run `msbuild /t:pack /v:m` in the repository root; otherwise, the CI build will fail.

- This command must be run from a Visual Studio Developer Command Prompt
- Alternatively, `dotnet msbuild RoslynAnalyzers.sln -t:pack -v:m` can be used from a standard terminal window

Note: Consider merging the PR base branch (`2.9.x`, `main`, or `release/*`) into your branch before you run the pack command to reduce merge conflicts.

-->

Kritner commented:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=639007&view=logs&j=f0f0520d-467b-5d1d-a95d-4e3a2581fe8a&t=c140c00f-6c0a-5356-d099-4e047a5525ed&l=221

![image](https://github.com/dotnet/roslyn-analyzers/assets/8814983/3c595425-c199-47fc-9edd-797066847d9a)

getting this build error, both prior to and after making updates to the files mentioned.  I'm not sure if this command is supposed to be run locally, or as a part of the build automatically; but i was getting errors attempting to run it locally.

going to revert changes to these two files mentioned and just double check the same error occurs:

```
One or more auto-generated documentation files were either edited manually, or not updated. Please revert changes made to the following files (if manually edited) and run `msbuild /t:pack` at the root of the repo to automatically update them:
      D:\a\_work\1\s\src\NetAnalyzers\Microsoft.CodeAnalysis.NetAnalyzers.md
      D:\a\_work\1\s\src\NetAnalyzers\Microsoft.CodeAnalysis.NetAnalyzers.sarif
```

Kritner commented:
@buyaa-n i think this might be good to go now - though i'm a bit unsure of the git flow being used.  I'm assuming i want to target my PR to the `release/8.0.2xx` branch such that it will get release under the next build of those revision of packages?  Or should this be targeting main?

(I only tagged you because of your activity on #7285, lmk if i should tag someone else or just let this thing sit til it gets through some queue)

tarekgh commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:
```diff
@@ -251,6 +268,45 @@ private void AnalyzeFormatArgument(OperationAnalysisContext context, IOperation
+        /// <returns>When true brackets are valid, false otherwise.</returns>
+        private static bool IsValidBraces(string text)
+        {
+            var stack = new Stack<char>();
```
This is not going to work well as the `{` and `}` can be escaped. Please look at https://source.dot.net/#Microsoft.Extensions.Logging.Abstractions/LogValuesFormatter.cs,42 how these are parsed. 

CC @stephentoub

Kritner replied:
heh... part of the problem is that `LogValuesFormatter` is what is *throwing* when passing in a bad format string from the original issue 😅 

could probably use some of the code in there as a way to bootstrap a working solution on the analyzer side - but the real problem is that now such badly formatted message templates throw with the stack trace:

```cs
   at System.ThrowHelper.ThrowFormatInvalidString(Int32 offset, ExceptionResource resource)
   at System.Text.CompositeFormat.Parse(String format)
   at Microsoft.Extensions.Logging.LogValuesFormatter..ctor(String format)
   at Microsoft.Extensions.Logging.FormattedLogValues.<>c.<.ctor>b__9_0(String f)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.Logging.FormattedLogValues..ctor(String format, Object[] values)
   at Microsoft.Extensions.Logging.LoggerExtensions.Log(ILogger logger, LogLevel logLevel, EventId eventId, Exception exception, String message, Object[] args)
   at Microsoft.Extensions.Logging.LoggerExtensions.Log(ILogger logger, LogLevel logLevel, String message, Object[] args)
   at Microsoft.Extensions.Logging.LoggerExtensions.LogInformation(ILogger logger, String message, Object[] args)
   at Program.<Main>$(String[] args) in C:\Users\kritn\source\repos\ConsoleApp1\ConsoleApp1\Program.cs:line 13
```

from a line like:

```cs
logger.LogInformation(@"Test: {i}{", i);
```

![image](https://github.com/dotnet/roslyn-analyzers/assets/8814983/192148dd-3eca-4ff6-8482-8159f70bf585)

tarekgh replied:
> but the real problem is that now such badly formatted message templates throw at runtime with the stack trace:

Can you clarify why this is a problem?

Kritner replied:
a problem in the general sense - as in the reasoning for submitting the issue. Additionally having been pointed to code from your comment https://github.com/dotnet/roslyn-analyzers/pull/7286#discussion_r1561388188 that actually *is* the issue that needs fixing (or flagging through the analyzer) to avoid the run time exceptions

tarekgh replied:
To be honest, I don't perceive it as a problem. The behavior, throwing an exception at runtime, is correct. Implementing an analyzer to detect the issue at compile time serves as a helpful means of catching it early, much like we do with other issues we create analyzers for.

Kritner replied:
hmm... i was under the impression this was new behavior - we had some (seemingly) hot path code start erroring out in prod after the .net8 update, where it wasn't in 7. unless this particular function that had a log statement just hadn't been hit in over a year.

If it's existing this way for a while though then yeah i guess it's not that big of a deal, just don't see how it couldn't be new behavior /shrug

tarekgh replied:
It was a bug which is fixed. The mal-formed message shouldn't be allowed and should throw as we do now in .NET 8.0. It was just a luck that this problem existed before and not noticed.

Kritner replied:
believe this is taken care of now through https://github.com/dotnet/roslyn-analyzers/pull/7286/commits/7333c4d2d7dc37d7b38c6f13ead7935da98d5a43

buyaa-n commented:
>  I'm assuming i want to target my PR to the release/8.0.2xx branch such that it will get release under the next build of those revision of packages? Or should this be targeting main?

No this should go to main, only servicing changes should go to release branches

buyaa-n commented on src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.md:
```diff
@@ -1944,9 +1944,9 @@ Forward the 'CancellationToken' parameter to methods to ensure the operation can
 ---
 
-## [CA2017](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2017): Parameter count mismatch
+## [CA2017](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2017): Parameter count mismatch or invalid braces in message template
```
Hm, I thought only additional message will be enough, not sure we could/should change a title of an analyzer that released in .NET 6.0, maybe we should use a new rule ID? What you think @stephentoub @sharwell @mavasani?

Kritner replied:
yeah i think that makes sense to handle it under a separate ID, it seems like from the localization generation "last in" wins, which had *completely* replaced the old title with the new. So i went the route of combining the title messages, keeping the "message" (or was it description?) distinct between the two descriptors

Kritner replied:
Introduced as new `CA2023` instead of overloading `CA2017`.  `CA2023` is specific to orphaned unescaped braces

buyaa-n replied:
Yes, new rule needs design review, though this is not like a completely new rule, so not sure, in case it needs API review we also need to move the issue back to runtime.

Kritner replied:
looks like i can add a few more tests to help get coverage back up, but lmk if there's anything else needed from me

tarekgh commented:
@stephentoub could you please advise if we can handle this in already existing rule or do you think we need to introduce a new analysis rule with new diagnostic?

tarekgh commented:
@Kritner thanks for helping with this issue. 

If we'll continue having a new diagnostic we'll need to get approval for it before we proceed. Could you please log a new issue in the runtime repo to track reviewing it? here is some example issue https://github.com/dotnet/runtime/issues/78406 which you can mimic. meanwhile I converted this PR to draft till we finalize the process.

Kritner commented:
from @tarekgh: 

> If we'll continue having a new diagnostic we'll need to get approval for it before we proceed. Could you please log a new issue in the runtime repo to track reviewing it? here is some example issue https://github.com/dotnet/runtime/issues/78406 which you can mimic. meanwhile I converted this PR to draft till we finalize the process.

added https://github.com/dotnet/runtime/issues/101698

tarekgh commented on src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.md:
```diff
@@ -2070,6 +2070,18 @@ A call to 'Stream.Read' may return fewer bytes than requested, resulting in unre
 
+## [CA2023](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2023): Invalid braces in message template
+
+The braces present in the message template are invalid.  Ensure any braces in the message template are valid opening/closing braces, or are escaped.
```
>  [](http://example.com/codeflow?start=56&length=1)

nit: remove the extra space.

tarekgh replied:
Yes, I meant the extra space after the `.`. Sorry for not been clear.

tarekgh commented on src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.md:
```diff
@@ -2070,6 +2070,18 @@ A call to 'Stream.Read' may return fewer bytes than requested, resulting in unre
+|-|-|
+|Category|Reliability|
+|Enabled|True|
+|Severity|Error|
```
>Error [](http://example.com/codeflow?start=10&length=5)

In the design review it is stated we need to have the same severity as CA2017 which is `warning`

tarekgh commented on src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.sarif:
```diff
@@ -3711,6 +3711,26 @@
+        "CA2023": {
+          "id": "CA2023",
+          "shortDescription": "Invalid braces in message template",
+          "fullDescription": "The braces present in the message template are invalid.  Ensure any braces in the message template are valid opening/closing braces, or are escaped.",
```
>  [](http://example.com/codeflow?start=86&length=1)

remove the extra space.

tarekgh commented on src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.sarif:
```diff
@@ -3711,6 +3711,26 @@
+          "id": "CA2023",
+          "shortDescription": "Invalid braces in message template",
+          "fullDescription": "The braces present in the message template are invalid.  Ensure any braces in the message template are valid opening/closing braces, or are escaped.",
+          "defaultLevel": "error",
```
>error [](http://example.com/codeflow?start=27&length=5)

warning

tarekgh commented on src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md:
```diff
@@ -9,6 +9,7 @@ CA1515 | Maintainability | Disabled | MakeTypesInternal, [Documentation](https:/
 CA1871 | Performance | Info | DoNotPassNonNullableValueToArgumentNullExceptionThrowIfNull, [Documentation](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1871)
 CA1872 | Performance | Info | PreferConvertToHexStringOverBitConverterAnalyzer, [Documentation](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1872)
 CA2022 | Reliability | Warning | AvoidUnreliableStreamReadAnalyzer, [Documentation](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2022)
+CA2023 | Reliability | Error | LoggerMessageDefineAnalyzer, [Documentation](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2023)
```
>Error [](http://example.com/codeflow?start=23&length=5)

warning

tarekgh commented on src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx:
```diff
@@ -189,6 +189,15 @@
+    <value>The braces present in the message template are invalid</value>
+  </data>
+  <data name="LoggerMessageDiagnosticMessageTemplateBracesMismatchDescription" xml:space="preserve">
+    <value>The braces present in the message template are invalid.  Ensure any braces in the message template are valid opening/closing braces, or are escaped.</value>
```
>  [](http://example.com/codeflow?start=67&length=1)

remove the extra space

tarekgh commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:
```diff
@@ -82,7 +86,17 @@ public sealed class LoggerMessageDefineAnalyzer : DiagnosticAnalyzer
+                                                                         CreateLocalizableResourceString(nameof(LoggerMessageDiagnosticMessageTemplateBracesMismatchTitle)),
+                                                                         CreateLocalizableResourceString(nameof(LoggerMessageDiagnosticMessageTemplateBracesMismatchMessage)),
+                                                                         DiagnosticCategory.Reliability,
+                                                                         RuleLevel.BuildError,
```
>BuildError [](http://example.com/codeflow?start=83&length=10)

BuildWarning

tarekgh commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/LoggerMessageDefineAnalyzer.cs:
```diff
@@ -251,6 +271,135 @@ private void AnalyzeFormatArgument(OperationAnalysisContext context, IOperation
+        /// <param name="formatter">The text to check.</param>
+        /// <returns>When true braces are valid, false otherwise.</returns>
+        private static bool IsValidBraces(LogValuesFormatter formatter)
+        {
```
I think this parsing logic can be simplified without using place holders at all. Just by scanning the text one time. The code [https://source.dot.net/#Microsoft.Extensions.Logging.Abstractions/LogValuesFormatter.cs,88](https://source.dot.net/#Microsoft.Extensions.Logging.Abstractions/LogValuesFormatter.cs,88) can give some idea how we can do that. This code is not exact what we need but we can use similar idea. 

Let me know if you need any help writing it.

tarekgh replied:
Can you try/test the code:

```C#
        public static bool IsValidMessageTemplate(string messageTemplate)
        {
            if (messageTemplate is null)
            {
                return false;
            }

            if (messageTemplate.Length == 0)
            {
                return true;
            }

            int index = 0;
            bool leftBrace = false;

            while (index < messageTemplate.Length)
            {
                if (messageTemplate[index] == '{')
                {
                    if (index < messageTemplate.Length - 1 && messageTemplate[index + 1] == '{')
                    {
                        index++;
                    }
                    else if (leftBrace)
                    {
                        return false;
                    }
                    else
                    {
                        leftBrace = true;
                    }
                }
                else if (messageTemplate[index] == '}')
                {
                    if (leftBrace)
                    {
                        leftBrace = false;
                    }
                    else if (index < messageTemplate.Length - 1 && messageTemplate[index + 1] == '}')
                    {
                        index++;
                    }
                    else
                    {
                        return false;
                    }
                }

                index++;
            }

            return !leftBrace;
        }
```

Kritner replied:
https://github.com/dotnet/roslyn-analyzers/pull/7286#discussion_r1715996888 basically a leetcode problem 😅

Kritner commented:
Think i handled all the comments from the PR review, wasn't sure if i should be the one "resolving conversations" or if you'd be doing that @tarekgh.  Thanks for the help!

tarekgh reviewed (approved):
LGTM. Thanks @Kritner!

tarekgh commented:
@buyaa-n could you please have a quick look and merge if it looks good to you too?

buyaa-n commented:
> @buyaa-n could you please have a quick look and merge if it looks good to you too?

Is this analyzer needed for v9? I would wait until the main open for v10.

tarekgh commented:
@buyaa-n could you please track it to merge when the main open for v10 merges?

buyaa-n commented:
> @buyaa-n could you please track it to merge when the main open for v10 merges?

I plan to go through all PRs and review/merge when main opens for v10

Kritner commented:
Added a few more covering tests based on the new implementation

buyaa-n reviewed (approved):
Looks good, thank you @Kritner!

xtqqczze commented:
CA2023 is still missing documentation: https://github.com/dotnet/sdk/issues/51305
